### PR TITLE
attribute #[using] for Definition and Fixpoint

### PR DESCRIPTION
**Pull request by gares**
My original need was to have this facility at the ML API level (for Elpi).
While I was at it I've also exposed the feature to the concrete syntax via the attributes mechanism.

### CODE REVIEW

**Comment by gares** [comment]
CC @CohenCyril @pi8027

File: vernac/g_vernac.mlg
Comment (SkySkimmer) [inline vernac/g_vernac.mlg:117]: Can parser experts make a general `ident_or_keyword` entry instead of hardcoding it here?
@coq/parsing-maintainers

Comment (gares) [reply vernac/g_vernac.mlg:117]: that would be nice indeed

**Comment by Zimmi48** [comment]
I can't help but notice that all the examples have `using="All"`.

**Comment by gares** [comment]
Hum, the point was to test that the attribute is actually taken into account by all (hopefully) code paths. And it seems it is.
I can add more, after all they won't hurt.

**Comment by gares** [comment]
@Zimmi48 I've added more tests and a changelog entry

File: test-suite/success/definition_using.v
Comment (gares) [inline test-suite/success/definition_using.v:26]: CC @palmskog this is just a heads up. Even if the PR is talking about Definition/Fixpoint in non-interactive mode, the attribute also works for interactive mode. So one can use it in place of `Proof using`. I know that in Coq community you are advising to be explicit about the section fragment being used, thanks! You may now consider also using an attribute. (I personally never liked to annotate `Proof` with the variables, but back then attributes did not exist).

**Comment by gares** [comment]
Dear @coq/doc-maintainers I need to describe this in the doc. One obvious place is https://coq.inria.fr/refman/proof-engine/proof-handling.html#coq:cmd.proof-using as an alternative syntax for `Proof using`, another place is https://coq.inria.fr/refman/language/core/definitions.html where we list the attributes supported by the commands described in this section. Would that work? Any better ideas?

**Comment by gares** [comment]
I love the linter: `No newline at end of file ./doc/changelog/02-specification-language/13183-using-att.rst!`. If only it was *not* a generated file... #13187

File: test-suite/success/definition_using.v
Comment (herbelin) [inline test-suite/success/definition_using.v:10]: Is there a rationale regarding when to use `foo="value"` or `foo(value)`, as in e.g. `#[universes(template)]`?

Similarly, is there a rationale regarding `using="ccc"` vs say `using(collection(ccc))`, or of `using="a*"` vs, say, `using(anyextensionof(a))`?

**Comment by Zimmi48** [comment]
> Dear @coq/doc-maintainers I need to describe this in the doc. One obvious place is https://coq.inria.fr/refman/proof-engine/proof-handling.html#coq:cmd.proof-using as an alternative syntax for `Proof using`, another place is https://coq.inria.fr/refman/language/core/definitions.html where we list the attributes supported by the commands described in this section. Would that work? Any better ideas?

I would mention the attribute in the list present in the description of `Definition` and `Fixpoint` but put the presentation of the attribute next to the `Proof using` syntax so that the documentation can be shared as much as possible.

File: doc/changelog/02-specification-language/13183-using-att.rst
Comment (Zimmi48) [inline doc/changelog/02-specification-language/13183-using-att.rst:3]:
```suggestion
  Attribute :attr:`using` now supported by Definition and (Co)Fixpoint.
  It has the same effect as :cmd:`Proof using`, which is only available in
```

(Note the top part of my suggestion requires adding the documentation of the attribute, with the `.. attr::` directive first; while the bottom part is needed to make the refman compilation succeed.)

File: test-suite/success/definition_using.v
Comment (gares) [reply test-suite/success/definition_using.v:10]: Well, one is that the syntax is the same of `Proof using` (same feature, same syntax).
Sure, one needs "quotes" here, but that is all.
Also, the AST of a collection is not trivial, encoding it is "function application" form would be hard to read, e.g. no infix.

Comment (gares) [reply test-suite/success/definition_using.v:10]: My favorite would be `#[using(ccc - h)]` but this is not possible as far as I can tell.

Comment (herbelin) [reply test-suite/success/definition_using.v:10]: But then, if the syntax of a collection is not trivial, are attributes the right tool for it? The idea with attributes was precisely that they did not come with a complex syntax.

Comment (gares) [reply test-suite/success/definition_using.v:10]: I don't follow, `#[using="collection"]` is simple and consistent with `Proof using collection`.
What is not trivial is encoding it into a rose.

**Comment by Zimmi48** [comment]
BTW, if you are not in a hurry, wait for #12936 to be in, as this PR changes the documentation of `Proof using`.

**Comment by gares** [comment]
I wrote some doc. Then feel free to merge the other, larger, PR before this one.

File: doc/sphinx/proof-engine/proof-handling.rst
Comment (Zimmi48) [inline doc/sphinx/proof-engine/proof-handling.rst:213]:
```suggestion
      a :n:`@collection`, between quotes, as its value. The effect is
      the same that one obtains by specifying the same :n:`@collection` to
```

**Review by jfehrle — DISMISSED** [review]
Some suggestions.

**Comment by gares** [comment]
Thanks for the hints. I'll apply the indentation change for sure. I think I did mention this attribute in the doc about Definition & co, there was already a paragraph about attributes. Is it ok or you want me to mention it elsewhere?

**Comment by jfehrle** [comment]
Yes, you already mention the attribute for `Definition`.  Would you also mention it in `Fixpoint`, `CoFixpoint`, and `Lemma`?  Can be as simple as ``` This command accepts the :attr:`using` attribute. ```

**Comment by gares** [comment]
doc update and git history cleaned up

**Comment by gares** [comment]
This PR needs a review by @coq/vernac-maintainers I guess and an assignee (I guess it can be from @coq/doc-maintainers so that the merge can be coordinated with  #12936

**Comment by Zimmi48** [comment]
PR #12936 has been merged. Can you rebase?

@coq/vernac-maintainers: We still need a review from you.

File: vernac/vernacentries.ml
Comment (SkySkimmer) [inline vernac/vernacentries.ml:504]: Not sure if this is better, keep the current version if you prefer it.
```suggestion
  let initial_goals pf = Proofview.initial_goals Proof.((data pf).entry) in
```

File: vernac/attributes.ml
Comment (SkySkimmer) [inline vernac/attributes.ml:232]: Having the error not match the attribute has the potential for confusion imo.
```suggestion
  |  _ -> CErrors.user_err (Pp.str "Ill formed \"using\" attribute")
```

File: stm/stm.ml
Comment (SkySkimmer) [inline stm/stm.ml:2514]: Why is there a case for the attribute but not for when the `Proof using` command or default proof using occurred?
Can we change it so they have more similar code paths?

**Review by SkySkimmer — APPROVED** [review]

File: stm/stm.ml
Comment (gares) [reply stm/stm.ml:2280]: We could, but it is not trivial. The point is that `collect_proof` scans the AST for a `VernacProof...`. For the attribute one should do something similar but I don't even know if the AST I have here includes the attribute or not.

There is already this codepath that is taken when there is no "Proof using" annotation, and that forces the state at the begin of the proof. Once you have that, it is trivial to fetch the (result of the) attribute.

**Comment by gares** [comment]
done

File: vernac/g_vernac.mlg
Comment (herbelin) [reply vernac/g_vernac.mlg:117]: Regarding declaration of tokens, a direction I'm currently exploring is rather to have locally defined keywords, that is, each rule (or rule prefixes) declares the keywords specific to its.

**Comment by herbelin** [comment]
Beside exposing an ML API, is the purpose of the feature to be able to type a definition without having to interpret its body? If not and that it does not provide more than `Proof using`, is the only other purpose to provide an attribute syntax to `Proof using`?

If yes, I have questions about the syntax, but, to be honest, I did not look closely at the feature in the past, so, since you're asking for a review, is it ok if I'm also asking general questions about `Proof using`, since after all, these more general questions have an impact on the syntax design choices made in this PR?

For instance, do you have typical examples of files which use the `Collection`, `+` or `a*` notation? Grepping in a couple of user developments, I found `Proof using Type*` often but never `Collection` , `+` and `a*`. So, how to convince that those are worth being used?

Also, is it correct that `All` does no test anything at all? And, if yes, what are the situations where it is useful to have `Proof using All` rather than nothing?

Also, is there a way to tell that I want to use exactly the given variables and not only a subset of the given variables?

Thanks in advance for answers which I believe can help to understand better what we need.

**Comment by gares** [comment]
> For instance, do you have typical examples of files which use the `Collection`, `+` or `a*` notation? Grepping in a couple of user developments, I found `Proof using Type*` often but never `Collection` , `+` and `a*`. So, how to convince that those are worth being used?

I don't, it may be an unused/unknown feature. This PR does not change this situation, nor change the syntax of Collection.
One can even remove Collection and this PR would just be fine.

> Also, is it correct that `All` does no test anything at all? And, if yes, what are the situations where it is useful to have `Proof using All` rather than nothing?

I don't follow, one of the reasons of existence of Proof using is to be able to write `Definition foo .... Proof using something. Admitted.` and have Coq make foo depend on the desired set of variable. `All` picks them all, if you say nothing Coq picks only the ones in the statement or partial proof term.

> Also, is there a way to tell that I want to use exactly the given variables and not only a subset of the given variables?

No, and I don't see how this impact this PR.

**Comment by herbelin** [comment]
>  `All` picks them all, if you say nothing Coq picks only the ones in the statement or partial proof term.

Sorry, I had a misconception about `All`. And I also suspect a misunderstanding on whether nothing meant `Proof.` (which is like `Proof using All`, iiuc) or `Proof using.` (which is like `Proof using Type.`, iiuc).

Regarding the syntax, my point is that due to the `+`, the syntax of `Proof using` is a syntax of term with infix symbols which requires a special parser. On the other side, the syntax of attributes is basic and does not support infix symbols. To circumvent this, the PR encodes the syntax of `Proof using` in a string and uses its own parser to decode the string. Nothing bad per se, besides a bit of disappointment. However, if the language in use for `Proof using` is in practice simpler, and if only, say, `All`, `Type`, `Type*` and sequences of `id` are really needed, it would not be needed anymore to use a string encoding and the syntax could use subattributes of the form say `all`, `type`, `type*`, `names="H1 H2"` which seem to more faithfully follow the rationale of the syntax of attributes (as used e.g. in `universes(cumulative)`, `universes(monomorphic)`, etc.).

PS: Not expecting all declared variables to be effectively used in the proof looks actually also ok. Sorry for my time needed to understand.

**Comment by Zimmi48** [comment]
@herbelin `Proof.` is neither equivalent to `Proof using Type.` nor to `Proof using All.`

See this example:

```coq
Section foo.

  Variable A B : Prop.
  Hypothesis (a : A) (equiv : A <-> False).

  Lemma lemma1 : False.
  Proof.
  apply equiv, a.
  Qed.

  Lemma lemma2 : False.
  Proof using All.
  apply equiv, a.
  Qed.

  Lemma lemma3 : False.
  Proof using Type.
  apply equiv, a.
  Fail Qed.
  (* The following section variables are used but not declared:
  equiv a A. *)
  Abort.

End foo.

Check lemma1. (* forall A : Prop, A -> A <-> False -> False *)
Check lemma2. (* forall A : Prop, Prop -> A -> A <-> False -> False *)
```

**Comment by gares** [comment]
So you are proposing a special syntax for the most common cases, it makes sense.

The ones I see as implementable are:
```
#[using(all)]   
#[using(type)]   
#[using="..."]
```
I don't think the `*` thing is possible.

Do you want me to add the two missing cases?

**Comment by SkySkimmer** [comment]
`using(collection)` should work too no? with `all` and `type` being pseudo collections.

**Comment by gares** [comment]
> `using(collection)` should work too no? with `all` and `type` being pseudo collections.

I was thinking the same ;-)
The implementation of the other would be to elaborate `(all)` to `="All"` and `(type)` to `="Type"` so I guess it is doable.

**Comment by SkySkimmer** [comment]
>Maybe we should have #[using(All)] directly...

Yes, I think having the 2 casings in 2 different places would get confusing.

**Comment by SkySkimmer** [comment]
TBH I wouldn't do the special syntax, `using="Type"` works fine.

**Comment by gares** [comment]
OK, I think I agree with @SkySkimmer. The implementable specialized syntax would be too similar to the generic one to be worth it.

Look like metacoq failure is a race in the makefile, I've restarted it. Other than that CI looks OK.

This PR still lacks an assignee from @coq/vernac-maintainers

**Comment by Zimmi48** [comment]
> Look like metacoq failure is a race in the makefile, I've restarted it.

Yes, I've seen this happen very frequently. Maybe this should be reported to MetaCoq maintainers in case they are not already aware...

**Comment by herbelin** [comment]
> TBH I wouldn't do the special syntax, `using="Type"` works fine.

If at least @SkySkimmer and you (@gares) agree, I'm rallying to your choices.

**Comment by gares** [comment]
Can someone take this is PR?

**Comment by SkySkimmer** [comment]
I'll take it.

While you rebase, please squash the last commit with the first one (the "improve indentation" commit message is a bit incorrect btw).

**Comment by gares** [comment]
should be ok

**Comment by SkySkimmer** [comment]
coquelicot failure is spurious

@coqbot: merge now

**Comment by SkySkimmer** [comment]
@coqbot: merge now

**Comment by CohenCyril** [comment]
Hi! Sorry for coming in so late, but can someone point me to the introduction of collections and their doc? Can there be a collection with the same name as a definition? What if someone defines a collection or constant named "All"?

**Comment by gares** [comment]
The documentation is in the refman, just look for `Collection` in the command index: https://coq.inria.fr/refman/proof-engine/proof-handling.html#coq:cmd.collection

You cannot override the meaning of All and Type, but the system gives you no error, which is suboptimal.
```coq
Coq < Collection All := .

Coq < Section Foo
Coq < .

Coq < 
Coq < Variable x : nat.
x is declared

Coq < Lemma a : True.
1 subgoal
  
  x : nat
  ============================
  True

a < Proof using All.

a < trivial.
No more subgoals.

a < Qed.

Coq < End Foo.

Coq < About a.
a : nat -> True

a is not universe polymorphic
Arguments a _%nat_scope
a is opaque
Expands to: Constant Top.a
```

**Comment by Zimmi48** [comment]
@CohenCyril 
https://coq.github.io/doc/master/refman/proof-engine/proof-handling.html#name-a-set-of-section-hypotheses-for-proof-using

> Can there be a collection with the same name as a definition? What if someone defines a collection or constant named `All`?

No error is raised currently in any of the two cases (and BTW, you can also name a section variable `All`).

A collection name takes precedence over a section variable name according to my tests:

```coq
Section foo.
Variable A : Prop.
Collection A := Type.
Lemma foo : True.
Proof using A.
set (a := A).
easy.
Fail Qed.
```

But the same is not true for `All` (i.e. `All` is never interpreted as the collection name nor as the section variable):

```coq
Section foo.
Variable A : Prop.
Collection All := Type.
Lemma foo : True.
Proof using All.
set (a := A).
easy.
Qed.
```

```coq
Section foo.
Variable A All : Prop.
Lemma foo : True.
Proof using All.
set (a := A).
easy.
Qed.
```

**Comment by Zimmi48** [comment]
Oh, too bad GitHub didn't make the comment above appear. It would have avoided me to restate the same facts.

**Comment by CohenCyril** [comment]
I think this is wrong that there is no error or warning. And I did not see `Collection` be designed and merged. Was the name clashing discussed? The documentation says nothing about what would happen to e.g.
```coq
Section CollectionTest.
Variable (x y : nat).
Collection x := y.
#[using="x"] Definition foo := True.
End CollectionTest.
```

**Comment by CohenCyril** [comment]
I'd say either the namespace should be the same (i.e. defining a collection with the same name as a definition should raise an error), or they should be different and there should be a way to distinguish a collection from a name, as in `#[using=x]` for the collection and `#[using={x}]` for a single name.

**Comment by Zimmi48** [comment]
> And I did not see `Collection` be designed and merged

That's because it's a very old feature! It was introduced at the same time as `Proof using` (i.e. 8.5): https://coq.inria.fr/distrib/V8.5/refman/Reference-Manual009.html#hevea_command157

>Was the name clashing discussed?

I was not there, but I bet it wasn't, given the way things were done at that time.

>I think this is wrong that there is no error or warning.

I fully agree. Would you be so kind to open a new issue regarding this matter? (We are slightly hijacking a PR which is completely orthogonal to the `Collection` feature.)

**Comment by SkySkimmer** [comment]
Collection is pretty old, although it used to be called Package b49d803286ba9ed711313702bb4269c5e9c516fa ab7d96a1d7c1bd44cf3bca5f593b947975117ed8

**Comment by CohenCyril** [comment]
> I fully agree. Would you be so kind to open a new issue regarding this matter? (We are slightly hijacking a PR which is completely orthogonal to the `Collection` feature.)

I do not think it is, we can still make some things (more) right here...

**Comment by Zimmi48** [comment]
Also, when creating this issue, please use the existing syntax `Proof using` instead of the new attribute syntax like in your comment above:

> e.g., `#[using="c"]` for the collection and `#[using="{x}"]` for a single constant name (and `#[using="c + {x y}"]` for mixing).

Otherwise, you are making it look like the issue is related to the new feature.

**Comment by Zimmi48** [comment]
I don't know why GitHub stop showing new comments in real time to me :cry: 

>I do not think it is, we can still make some things (more) right here...

I disagree. There's no reason to change the syntax or the semantics in the `using` attribute and to leave the `Proof using` syntax / semantics unchanged. It's quite easy to change both.

**Comment by CohenCyril** [comment]
> Also, when creating this issue, please use the existing syntax `Proof using` instead of the new attribute syntax like in your comment above:
> 
> > e.g., `#[using="c"]` for the collection and `#[using="{x}"]` for a single constant name (and `#[using="c + {x y}"]` for mixing).
> 
> Otherwise, you are making it look like the issue is related to the new feature.

Sure, my point here being: fixing the a old `Proof using` syntax is not backward compatible, we can still fix the new `using` attribute and deprecate the old, making this right.

**Comment by gares** [comment]
FTR `Collection` was added just after `Proof using` as a mean to make `Proof using ...` lines shorter. The name of the command was the result of a poll among native speakers (hence the old name dig out by Gaetan).

I don't see a use case for naming both a variable and a collection with the same ident. IMO we should just error.

**Comment by gares** [comment]
>  (We are slightly hijacking a PR which is completely orthogonal to the `Collection` feature.)

I'll craft a new label: hijacked-postmortem

File: vernac/g_vernac.mlg
Comment (Zimmi48) [inline vernac/g_vernac.mlg:117]: If I understand correctly, this change was needed because `using` is defined as a keyword in the Ltac plugin.
In consequence, this makes `using` a keyword even in `-noinit` mode.
Anyway, I would have preferred a more robust change allowing the use of any keyword as an attribute name in the context of `#[ ... ]`. Wouldn't this have been possible?

Comment (herbelin) [reply vernac/g_vernac.mlg:117]: > a more robust change allowing the use of any keyword as an attribute name in the context of #[ ... ]. Wouldn't this have been possible?

Not without a significant change of the current (and standard) decoupling of the lexing/parser phases.

In the present case, it should work with `IDENT "using"` instead and this wouldn't declare a keyword until some other component does it.

Comment (Zimmi48) [reply vernac/g_vernac.mlg:117]: Good suggestion. Let's do this fix.

Comment (gares) [reply vernac/g_vernac.mlg:117]: Hum, isn't "using" always a keyword? I mean, maybe technically it declared in Ltac1's file, I don't know, but to me `Proof using` is vanilla Coq, not something you can unload with -noinit.

Comment (herbelin) [reply vernac/g_vernac.mlg:117]: This is the usual discussion, but, technically, the parser does not require the `using` of `Proof using` to be a keyword for the lexer. Then, of course, we may want to make it a reserved word independently of parsing reasons.

**Comment by gares** [comment]
I thought @Zimmi48 problem was with -no-init. On that I don't agree.
About making it a non-keyword everywhere, I'm fine, no preference, but it would be a change w.r.t. the status quo.

**Comment by Zimmi48** [comment]
I based myself on the manual to learn that `using` was only a keyword in the prelude and not in the initial state.
Then, following @gares's remark on `Proof using` (which I agree is fully part of the base vernacular), I did this test in `-noinit` mode:

```coq
Section A.
Variable A : Prop.
Hypothesis a : A.
Lemma b : A.
Proof using.
(* Error: The reference using was not found in the current environment. *)
```

This is a bug to me (it should not error).

File: vernac/g_vernac.mlg
Comment (Zimmi48) [reply vernac/g_vernac.mlg:117]: I was not raising "the usual discussion" about what should be keywords.

**Comment by gares** [comment]
`Proof term` is one thing, since you did not load ltac `using` is not a kwd hence it can be a term (name) which Coq could not find.
Welcome to the dark side.

**Comment by Zimmi48** [comment]
Yes, I know how to make sense of this error. But what I meant is that it is a bug that `Proof using` is not defined in this place:

https://github.com/coq/coq/blob/e5dc6e5cd189fb0e6fff672d7e978c62b4d4c160/vernac/g_proofs.mlg#L58-L60

And is defined here instead:

https://github.com/coq/coq/blob/e5dc6e5cd189fb0e6fff672d7e978c62b4d4c160/plugins/ltac/g_ltac.mlg#L330-L337

There's no valid justification for this.

**Comment by Zimmi48** [comment]
A bug that was introduced by @ppedrot (5466267afa78cac5479a503338fbc57d77f05458) in one of his insufficiently-reviewed pre-PR-era big code moves toward the creation of the Ltac plugin.

**Comment by gares** [comment]
Hum, this `Pltac` may be the reason. "unsing" could be moved, since its syntax is in vernac. But "with" needs the syntax of tactics which I guess is in Ltac.

**Comment by Zimmi48** [comment]
Would it break the parser if `Proof using` were in `g_proofs.mlg` and `Proof with` stayed where it is?

**Comment by herbelin** [comment]
You can try, but camlp5 should factorize automatically the shared "Proof", iianm.

It should also work with `IDENT "using"` instead of `"using"`.

**Pull request merged**
